shell.nix

```diff
@@ -5,13 +5,19 @@ pkgs.mkShell {
 
   packages = with pkgs; [
     cargo
+    rustc
     nodejs
     gcc
+    pnpm
+    git
 
     pkg-config
     cmake
     webkitgtk
     gnome.libsoup
+    librsvg
+    fmt_11
```

Comment:
There is a link error for choreo without it.
`/nix/store/x7yyxvwy1f9hlx72rzrgx069jyf7hxwr-binutils-2.42/bin/ld: cannot find -lfmt: No such file or directory`

Reply:
Then Nix is doing something wrong. TrajoptLib has the following in its build.rs:
```rs
    println!(
        "cargo:rustc-link-search=native={}/bin",
        cmake_dest.display()
    );
    println!(
        "cargo:rustc-link-search=native={}/lib",
        cmake_dest.display()
    );
    println!("cargo:rustc-link-lib=TrajoptLibRust");
    println!("cargo:rustc-link-lib=TrajoptLib");
    println!("cargo:rustc-link-lib=Sleipnir");
    println!("cargo:rustc-link-lib=fmt");
```
Maybe Nix is ignoring the linker search paths?

Linking against the system-provided fmt is not safe ABI-wise because it could be built with different compiler flags, or it could be a different version with an ABI break.

Reply:
Am I correct in guessing that the build succeeds on most systems because this `rustc-link-search` flag from the TrajoptLib build script is persistent throughout the other builds? Otherwise how is the final compilation step aware of the custom fmt build?

I should note that TrajoptLib builds perfectly fine, it's only the final compilation step of assembling all of the object files into the choreo binary that fails. When I put `-L./target/release/build/trajoptlib-f3ff2b1b047b8a2d/out/lib64/` into the `RUSTFLAGS` environment variable, the build succeeds. (although I would assume that just setting this variable in the shell.nix file is unfavorable due to the seemingly random string following build/trajoptlib- that, correct me if I'm wrong, may be subject to change depending on how that string is calculated)

If my earlier guess is correct, I think the shell provided by nix may be erasing some form of cache that stores the `rustc-link-search` option, (whether that be environment variables or something else) and I can try to see how to remedy this. If a solution that integrates nicely with the nix system is too difficult to achieve, is hard-coding the environment variable (if that string in the trajoptlib directory name can be determined simply) or maybe even building that specific version of fmt within the shell.nix declaration feasible? (although this is prone to falling behind in terms of versioning if the trajoptlib dependency is updated)

Reply:
> Am I correct in guessing that the build succeeds on most systems because this rustc-link-search flag from the TrajoptLib build script is persistent throughout the other builds? Otherwise how is the final compilation step aware of the custom fmt build?

TrajoptLibRust links statically to all its dependencies, so Choreo shouldn't link to fmt directly.
```bash
[tav@myriad Choreo]$ find . -type f | grep libfmt
./target/debug/build/trajoptlib-d24ea07aefebca26/out/build/_deps/sleipnir-build/libfmt.a
./target/debug/build/trajoptlib-d24ea07aefebca26/out/lib/libfmt.a
```

Reply:
Try as I may, I cannot figure out what could possibly be causing a stray `-lfmt` to appear in the final compilation step. I've built successfully in plain Arch Linux, in a pure nix shell underneath Arch Linux, and even in a pure, version locked nix flake underneath Arch Linux. Using any approach underneath my NixOS installation fails with this stray fmt linkage. (even when using the flake that produced success on Arch, which _should_ produce an environment exactly identical, including package versions, environment variables, etc.) I also tested to make sure that a program linking to fmt from within a nix environment under Arch Linux fails, so it's likely the command being run is different. Is there a log for what commands get executed during build?

Even in a fresh NixOS VM with the default configuration, `-lfmt` is still part of the compilation flags for Choreo, so the problem is likely common for any NixOS install.

Reply:
`cargo -v` or `cargo -vv` may get you the information you need. You could also do the following if C++ isn't giving enough information:
```diff
diff --git a/trajoptlib/build.rs b/trajoptlib/build.rs
index b952041..0a36c73 100644
--- a/trajoptlib/build.rs
+++ b/trajoptlib/build.rs
@@ -5,7 +5,8 @@ fn main() {
 
     cmake_config
         .profile("Release")
-        .define("BUILD_TESTING", "OFF");
+        .define("BUILD_TESTING", "OFF")
+        .define("VERBOSE", "ON");
 
     cmake_config.define("BUILD_SHARED_LIBS", "OFF");
```